crates/bevy_ui/src/render/ui_material_pipeline.rs

```diff
@@ -0,0 +1,737 @@
+/// [`UiMaterial`] asset type (which includes [`UiMaterial`] types).
+pub struct UiMaterialPlugin<M: UiMaterial>(PhantomData<M>);
+
+impl<M: UiMaterial> Default for UiMaterialPlugin<M> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+impl<M: UiMaterial> Plugin for UiMaterialPlugin<M>
+where
+    M::Data: PartialEq + Eq + Hash + Clone,
+{
+    fn build(&self, app: &mut bevy_app::App) {
+        load_internal_asset!(
+            app,
+            UI_VERTEX_OUTPUT_SHADER_HANDLE,
+            "ui_vertex_output.wgsl",
+            Shader::from_wgsl
+        );
+        load_internal_asset!(
+            app,
+            MATERIAL_UI_SHADER_HANDLE,
+            "ui_material.wgsl",
+            Shader::from_wgsl
+        );
+        app.init_asset::<M>()
+            .add_plugins(ExtractComponentPlugin::<Handle<M>>::extract_visible());
+
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
+            render_app
+                .add_render_command::<TransparentUi, DrawUiMaterial<M>>()
+                .init_resource::<ExtractedUiMaterials<M>>()
+                .init_resource::<ExtractedUiMaterialNodes<M>>()
+                .init_resource::<RenderUiMaterials<M>>()
+                .init_resource::<UiMaterialMeta<M>>()
+                .init_resource::<SpecializedRenderPipelines<UiMaterialPipeline<M>>>()
+                .add_systems(
+                    ExtractSchedule,
+                    (
+                        extract_ui_materials::<M>,
+                        extract_ui_material_nodes::<M>.in_set(RenderUiSystem::ExtractNode),
+                    ),
+                )
+                .add_systems(
+                    Render,
+                    (
+                        prepare_ui_materials::<M>.in_set(RenderSet::PrepareAssets),
+                        queue_ui_material_nodes::<M>.in_set(RenderSet::Queue),
+                        prepare_uimaterial_nodes::<M>.in_set(RenderSet::PrepareBindGroups),
+                    ),
+                );
+        }
+    }
+
+    fn finish(&self, app: &mut App) {
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
+            render_app.init_resource::<UiMaterialPipeline<M>>();
+        }
+    }
+}
+
+#[derive(Resource)]
+pub struct UiMaterialMeta<M: UiMaterial> {
+    vertices: BufferVec<UiMaterialVertex>,
+    view_bind_group: Option<BindGroup>,
+    marker: PhantomData<M>,
+}
+
+impl<M: UiMaterial> Default for UiMaterialMeta<M> {
+    fn default() -> Self {
+        Self {
+            vertices: BufferVec::new(BufferUsages::VERTEX),
+            view_bind_group: Default::default(),
+            marker: PhantomData,
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Pod, Zeroable)]
+pub struct UiMaterialVertex {
+    pub position: [f32; 3],
+    pub uv: [f32; 2],
+}
+
+/// in this [`UiMaterialPipeline`] there is (currently) no batching going on.
+/// Therefore the [`UiMaterialBatch`] is more akin to a draw call.
```

Comment:
```suggestion
// in this [`UiMaterialPipeline`] there is (currently) no batching going on.
// Therefore the [`UiMaterialBatch`] is more akin to a draw call.
```
This seems to be more a code comment than a doc